### PR TITLE
[New3D] TRIANGLE_LIST Markers

### PR DESCRIPTION
**Pull request by jhurliman**
**User-Facing Changes**

None

**Description**

Adds support for rendering TRIANGLE_LIST markers (user-supplied triangle vertices and colors). No EdgesGeometry for now since this can be slow to compute for a long triangle list that is changing frequently.

This also changes the React flow in ThreeDeeRender to avoid redundant React updates and/or multiple calls to `renderer.animationFrame()` in a single update.

### CODE REVIEW

File: packages/studio-base/src/panels/ThreeDeeRender/ThreeDeeRender.tsx [hunk 1]
```diff
@@ -415,8 +410,20 @@ export function ThreeDeeRender({ context }: { context: PanelExtensionContext }):
...
+    renderer?.setCameraState(cameraState);
+    renderRef.current.needsRender = true;
+  }, [cameraState, renderer]);
+
+  // Render a new frame if requested
+  if (renderer && renderRef.current.needsRender) {
     renderer.animationFrame();
-  }, [cameraState, messages, renderer, topicsToDatatypes]);
+    renderRef.current.needsRender = false;
+  }
```
Comment (jtbandes) [inline packages/studio-base/src/panels/ThreeDeeRender/ThreeDeeRender.tsx:426]: This should probably move into the useEffect right before renderDone

File: packages/studio-base/src/panels/ThreeDeeRender/renderables/markers/RenderableTriangleList.ts [hunk 2]
```diff
@@ -2,5 +2,139 @@
...
+    const prevMarker = this.userData.marker;
+    super.update(marker);
+
+    const vertexCount = marker.points.length;
+    const count = vertexCount * 3;
+    if (vertexCount === 0) {
+      this._renderer.layerErrors.addToTopic(
+        this.userData.topic,
+        NOT_DIVISIBLE_ERR,
+        `TRIANGLE_LIST: points[0] is empty`,
```
Comment (jtbandes) [inline packages/studio-base/src/panels/ThreeDeeRender/renderables/markers/RenderableTriangleList.ts:64]:
```suggestion
        `TRIANGLE_LIST: points is empty`,
```

File: packages/studio-base/src/panels/ThreeDeeRender/renderables/markers/RenderableTriangleList.ts [hunk 3]
```diff
@@ -2,5 +2,139 @@
...
+      );
+      this.geometry.setAttribute("position", new THREE.BufferAttribute(EMPTY_FLOAT32, 3));
+      return;
+    }
+    if (vertexCount % 3 !== 0) {
+      this._renderer.layerErrors.addToTopic(
+        this.userData.topic,
+        EMPTY_ERR,
+        `TRIANGLE_LIST: points[${vertexCount}] is not divisible by 3`,
```
Comment (jtbandes) [inline packages/studio-base/src/panels/ThreeDeeRender/renderables/markers/RenderableTriangleList.ts:73]: Should this be just a warning and we render up to `floor(vertexCount / 3)` vertices?

File: packages/studio-base/src/panels/ThreeDeeRender/renderables/markers/RenderableTriangleList.ts [hunk 4]
```diff
@@ -2,5 +2,139 @@
...
+      colors[i * 4 + 1] = tempColor.g;
+      colors[i * 4 + 2] = tempColor.b;
+      colors[i * 4 + 3] = tempColor.a;
+    }
+
+    if (dataChanged) {
+      this.geometry.setAttribute("position", new THREE.BufferAttribute(vertices, 3));
+      this.geometry.setAttribute("color", new THREE.BufferAttribute(colors, 4));
+      this.geometry.attributes.position!.needsUpdate = true;
+      this.geometry.attributes.color!.needsUpdate = true;
```
Comment (jtbandes) [inline packages/studio-base/src/panels/ThreeDeeRender/renderables/markers/RenderableTriangleList.ts:135]: what does this do? In the docs I see needsUpdate on Material, but these aren't materials, right?

**Review by jtbandes — APPROVED** [review]
I imagine this is going to require some significant changes to support triangle strip & triangle fan, but that can come later

File: packages/studio-base/src/panels/ThreeDeeRender/renderables/markers/RenderableTriangleList.ts (hunk 4 above)
Comment (jhurliman) [reply packages/studio-base/src/panels/ThreeDeeRender/renderables/markers/RenderableTriangleList.ts:135]: These buffers are the CPU-based mirror of buffers that also exist on the GPU. This tells 3js to sync CPU->GPU. I think they could already be set by the previous method calls, but those kind of internal details seem to change in 3js from version to version.

**Comment by jhurliman** [comment]
> I imagine this is going to require some significant changes to support triangle strip & triangle fan, but that can come later

My opinion is that we shouldn’t support these. 3js dropped support for them a while ago and just provides util methods to convert to triangle lists. There would be some bandwidth savings (not to the GPU, just disk or network) but compression would eliminate a lot of that savings.

File: packages/studio-base/src/panels/ThreeDeeRender/ThreeDeeRender.tsx (hunk 1 above)
Comment (jhurliman) [reply packages/studio-base/src/panels/ThreeDeeRender/ThreeDeeRender.tsx:426]: After making this change, camera input does not cause redraws while the player is paused.

**Pull request merged**